fuse/fuse_test.go

```diff
@@ -480,3 +480,99 @@ func TestWithRoot(t *testing.T) {
 		require.NotNil(t, err)
 	})
 }
+
+// Benchmarks
+
+var (
+	BenchmarkDataSizes = []int64{
+		0,
+		1024, 2*1024, 16*1024, 64*1024, 128*1023,
+		1*1024*1024, 16*1024*1024,
+	}
+)
+
+func stageAndRead(b *testing.B, ctx context.Context, control *spawntest.Control, mount *mountInfo, label string, data []byte) {
+	size := len(data)
+	// stage data to catFS
+	catfsFilePath := fmt.Sprintf("%s_file_%d", label, size)
+	req := catfsPayload{Path: catfsFilePath, Data: data}
+	require.Nil(b, control.JSON("/catfsStage").Call(ctx, req, &nothing{}))
+	fuseFilePath := filepath.Join(mount.Dir, catfsFilePath)
+
+	// Read it back via fuse
+	b.Run(fmt.Sprintf("%s_Size_%d", label, size), func(b *testing.B) {
+		for n := 0; n < b.N; n++ {
+			ioutil.ReadFile(fuseFilePath)
+		}
+	})
+}
+
+func BenchmarkRead(b *testing.B) {
+	withMount(b, MountOptions{}, func(ctx context.Context, control *spawntest.Control, mount *mountInfo) {
+		for _, size := range BenchmarkDataSizes {
+			// Check how fast is readout of a file with compressible content
+			data := testutil.CreateDummyBuf(size)
+			stageAndRead(b, ctx, control, mount,  "CompressibleContent", data)
+
+			// Check how fast is readout of a file with random/uncompressible content
+			data = testutil.CreateRandomDummyBuf(size, 1)
+			stageAndRead(b, ctx, control, mount,  "RandomContent", data)
+		}
+	})
+}
+
+func writeDataNtimes(b *testing.B, data []byte, ntimes int) {
+	// Writing could be very space demanding even for a small size,
+	// Since benchmark runs many-many times it will consume a lot of space
+	// we have to remount everything every time to start with clean catFS DB
+	// Consequently, this test takes long time, since mounting is long operation
+	label := "dummy"
+	size := len(data)
+	for n := 0; n < b.N; n++ {
+		b.StopTimer()
+		withMount(b, MountOptions{}, func(ctx context.Context, control *spawntest.Control, mount *mountInfo) {
+			// Check how fast is write to a file with compressible content
+			catfsFilePath := fmt.Sprintf("%s_file_%d", label, size)
+			fuseFilePath := filepath.Join(mount.Dir, catfsFilePath)
+
+			b.StartTimer()
+			for i := 0; i < ntimes; i++ {
+				if len(data)>0 {
+					// generate new content for backend
+					data[0]++
```

Comment:
Really minor nitpick: If ``ntimes`` would go over 256 easily, you could also use ``binary.LittleEndian.PutUint64()``  to write the full number to the first few bytes.

Reply:
Not sure that I understand. Why would `ntimes` go over 256. It is a constant in this code.

As for stamping the data. It content really is not important, as long as, at least one byte changes to produce a new `data` hash.

Reply:
> Not sure that I understand. Why would ntimes go over 256. It is a constant in this code.

It does not in the current code, but always imagine someone is changing the code that doesn't know it yet.
This someone could easily decide to add it to ``NumberOfOverWrites`` and expect that everything works.
It's just defensive programming and making code future proof for changes. After all ``ntimes`` is passed as ``int``.
> As for stamping the data. It content really is not important, as long as, at least one byte changes to produce a new data hash.

Yep, got that. That's why I got curious why only the first byte was written and what the edge cases are.

Reply:
> > Not sure that I understand. Why would ntimes go over 256. It is a constant in this code.
> 
> It does not in the current code, but always imagine someone is changing the code that doesn't know it yet.
> This someone could easily decide to add it to `NumberOfOverWrites` and expect that everything works.
> It's just defensive programming and making code future proof for changes. After all `ntimes` is passed as `int`.

Now I am confused, why we are worried about `256`? Neither `i` nor `ntimes` are used inside the loop
```
for i := 0; i < ntimes; i++ {
	if len(data) > 0 {
		// modification of one byte is enough
		// to generate new encrypted content for the backend
		data[0]++
	}
	require.NoError(b, ioutil.WriteFile(fuseFilePath, data, 0644))
}
```
`i:=0` is untyped but defaults to `int` (32 bits by go standard) which is the same as `ntimes`, so there should be no overflow. Unless some one puts a negative number for `ntimes` we should be OK. 

I added protection against negative input in ae4ba980

Reply:
It's not used inside the loop, but you increment ``data[0]`` ``ntimes``. If ``ntimes > 255`` the byte overflows and you will get a hash that you had before. This will be (might be?) faster than truly unique data and could influence the benchmark results. Therefore my comment about encoding ``i`` completely in the first few bytes with ``binary.LittleEndian.PutUint64()``.

Sure, it's a little contrived example considering that this won't happen in the test (``ntimes`` is always low), I just saw a opportunity for some general advice. Hope you don't mind.

Reply:
Now I get it. You are precisely correct and I am looking at wrong place.
Fixed with 65382764